### ISSUE TITLE
pandoc-run-pandoc doesn't seem to consume input

### ISSUE THREAD

**Issue by anghyflawn** [joostkremers/pandoc-mode#44]
I'd been using pandoc-mode with no issues but now it seems to have stopped consuming the input correctly. Everything works until the moment I try to run it. The output buffer shows the relevant pandoc call and a long string of ^G characters, but nothing happens. It's not a hang: the pandoc process is running and can be killed from outside Emacs, at which point the buffer correctly shows "Error in pandoc process".

I'm using the latest pandoc-mode from git on OS X 10.8 with Emacs 24.5. (Although the behaviour was also present in 24.3 which I was using until quite recently.) Any idea where I could start debugging this?

Thanks!

**Comment by joostkremers** [comment]
> Any idea where I could start debugging this?

Well, the first thing to try is whether the problem also occurs with `emacs -Q` followed by loading `pandoc-mode` (and preferably *only* `pandoc-mode`). If so, we can rule out your personal Emacs configuration. Then the question would be if it happens with any input / output format or just particular ones?

In principle, anything that gets put in the output buffer is coming from the subprocess, pandoc in this case, but I wouldn't know why pandoc would produce such a string of ^G characters. But then again, I wouldn't know why Emacs or pandoc-mode would produce them... Does this string keep growing?

Is there any pandoc-mode version that doesn't have the problem? The last commit that made changes to the code calling pandoc was 8f9be49 `Use an asynchronous process to avoid freezing Emacs`, so as a start, you could try commit 5c688ae right before it.

Also, running the function `pandoc--call-external` through edebug may provide some information.

**Comment by anghyflawn** [comment]
Sorry, I should have made it clear that I did try the `emacs -Q` route to make sure it's not an issue at my end. The problem is indeed absent in 5c688ae (both with `emacs -Q` and with my configuration).

The strings of bell characters does not grow. Stepping through `pandoc--call-external` with edebug doesn't throw up anything untoward as far as I can see --- it send the correct region to the process and then sends EOF once. Since the external process stays alive, might it have something to do with `process-send-eof` failing to do its job?

**Comment by joostkremers** [comment]
> Since the external process stays alive, might it have something to do with `process-send-eof` failing to do its job?

Possibly, because AFAIK Pandoc should terminate when it gets an eof.

That’s to say, I just noticed on the command line that pandoc doesn’t terminate on ^D when it’s not immediately preceded by a newline.

Which raises the question, does your text file have a final newline? If not, what happens if you add it? (Probably a long shot, but I really don’t have a clue… I’ll have to ask the Pandoc maintainers if there are any circumstances in which Pandoc might output a series of bell characters. Though I suspect that’s a long shot, too…)

**Comment by joostkremers** [comment]
BTW, have you tried piping your file to `pandoc` on the command line?

**Comment by anghyflawn** [comment]
Thanks! Yes, there is a newline and yes, piping the file to pandoc works correctly. Perhaps as a workaround the synchronous way of running pandoc could be made available again as an option? I can probably do it myself if you think it might be worth including.

**Comment by anghyflawn** [comment]
So [this](https://github.com/anghyflawn/pandoc-mode/commit/e7f6a13c7b47ae0117206b01a2b0d6a147f87fa3) seems to work -- I can't see anything breaking (i.e. I still get the string of bells) but setting `pandoc-use-async` to NIL gives me the old behaviour which does freeze up but at least work.

**Comment by joostkremers** [comment]
Yes, I had the same idea. I'll merge your code, thanks!

**Comment by ifitzpat** [comment]
I had this same issue with the latest commit. As @anghyflawn suggested `(setq pandoc-use-async nil)` fixed the problem.

**Comment by raamdev** [comment]
I also had just had this issue. I tried converting a buffer and just got "Running pandoc" indefinitely. It just looked like it was taking a really long time to do the conversion. Checking `ps | grep pandoc` revealed that the pandoc process was just sitting there, hung.

`(setq pandoc-use-async nil)` fixed the issue and the conversions now work as expected.

I'm using `pandoc-mode 20160405.2349` and `GNU Emacs 24.5.1`

**Comment by joostkremers** [comment]
Thanks for the report. I really don't know what it is that's going wrong, though... Since you're using `ps` and `grep`, I assume you're on Linux or OS X?

I'll see if I can find out what exactly could be done to debug this.

**Comment by raamdev** [comment]
Yes, I'm using OS X 10.11.4 (El Capitan) with Emacs compiled via Homebrew (see [`emacs.rb` formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/emacs.rb)), which built Emacs 24.5.1 from source (via http://ftpmirror.gnu.org/emacs/emacs-24.5.tar.xz) using `--with-cocoa --with-gnutls`.

**Comment by natecraig** [comment]
Changing the pandoc process connection type to a pipe resolved this issue for me.

**Comment by joostkremers** [comment]
I've made the connection type a user-configurable option.

Note that one other possible cause for the issue discussed here could be that the buffer being sent to pandoc doesn't have a final new line. I ran into that myself recently and it caused the pandoc process to just sit there and wait. Adding the final newline solved the issue. This may or may not be the cause of the problem for other people, but I thought it would be worth mentioning it here.

I'll close this issue now. With several solutions offered, there seems to be no need to keep it open.
